generator/src/main/java/com/graphicsfuzz/generator/fuzzer/OpaqueExpressionGenerator.java

```diff
@@ -1418,6 +1421,91 @@ public boolean preconditionHolds(Expr expr, BasicType basicType, boolean constCo
     }
   }
 
+  /**
+   * Identity transformation to insert a vector or matrix into a larger vector or matrix using a
+   * type constructor, then extract it back out again using another type constructor. When
+   * performed, transforms an expression e with type t -> identity(t(identity(m(identity(e), ..)))),
+   * where m is a type of equal or greater width than t.
+   * The rules for this smaller -> larger -> smaller transformation are as follows:
+   *     If the given type is a vector, the inner constructor type can be a vector of the same
+   *     element type with equal/greater width than the given type, or (provided that the given type
+   *     is a floating point genType) it can be a matrix with equal/greater column width than the
+   *     given type.
+   *     If the given type is a matrix, the inner constructor type can be a matrix with
+   *     equal/greater column/row width than the given type.
+   */
+  private class IdentityCompositeConstructorExpansion extends AbstractIdentityTransformation {
+    private IdentityCompositeConstructorExpansion() {
+      super(BasicType.allNumericTypes().stream().filter(
+          item -> !item.isScalar()).collect(Collectors.toList()), true);
```

Comment:
It seems limiting to disallow this for bool and bvec types - we'd like to be able to push a bool into a bvec3, or a bvec3 into a bvec4 - so long as the shading language supports boolean vector types.  So perhaps add bool and bvec types here if the shading language version allows it.

Also: maybe we do not want to allow types that cannot be put into something larger; i.e. not ivec4, uvec4, bvec4 and mat4x4.

Reply:
I did have boolean types enabled when working on this, but enabling them triggers an assertion failure in makeLiteralZeroOrOne():

`assert type.getElementType() != BasicType.BOOL;`

which seems a bit strange. From the context of the calling function (makeOpaqueZeroOrOne()), it seems like if the type is a boolean, then we should never be too deep (and so we should never have to make a literal zero or one) - but I'm not sure why that would be.